controllers/appgroup_controller.go

```diff
@@ -526,29 +526,10 @@ func (r *ApplicationGroupReconciler) cleanupWorkflow(ctx context.Context, logr l
 		for _, node := range wf.Status.Nodes {
 			nodes[node.ID] = node
 		}
-		graph, err := workflow.Build(g.Name, nodes)
+		// Generate the reversal Workflow object to submit to Argo
+		err := r.generateReversalWorkflow(ctx, logr, nodes, &g)
 		if err != nil {
-			logr.Error(err, "failed to build the wf status DAG")
-		}
-
-		rev := graph.Reverse()
-
-		for _, bucket := range rev {
-			for _, hr := range bucket {
-				err = pkg.HelmUninstall(hr.Spec.ReleaseName, hr.Spec.TargetNamespace)
-				if err != nil {
-					logr.Error(err, "failed to uninstall helm release using helm actions - continuing with cleanup")
-				}
-			}
-		}
-
-		for _, bucket := range rev {
-			for _, hr := range bucket {
-				err = r.Client.Delete(ctx, &hr)
-				if err != nil {
-					logr.Error(err, "failed to delete helmrelease CRO - continuing with cleanup")
-				}
-			}
+			logr.Error(err, "failed to generate reversal workflow")
```

Comment:
Hmm wondering if we should continue to brute force and carry on or actually return the error. I feel the latter might be better.

Reply:
Do you mean not calling r.Client.Delete if r.generateReversalWorkflow returned error?

Reply:
Now I return error if generateReverseWorkflow fails. It is probably the right thing to do because there is no reason for a requeue. Also added requeue logic. I will be testing this.